### PR TITLE
8254072: AArch64: Get rid of --disable-warnings-as-errors on Windows+ARM64 build

### PR DESCRIPTION
**Pull request by lewurm**
I organized this PR so that each commit contains the warning emitted by MSVC as commit message and its relevant fix.

Verified on
* Linux+ARM64: `{hotspot,jdk,langtools}:tier1`, no failures.
* Windows+ARM64: `{hotspot,jdk,langtools}:tier1`, no (new) failures.
* internal macOS+ARM64 port: build without `--disable-warnings-as-errors` still works. Just mentioning this here, because it's yet another toolchain (Xcode / clang) that needs to be kept happy [going forward](https://openjdk.java.net/jeps/391).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8254072](https://bugs.openjdk.java.net/browse/JDK-8254072): AArch64: Get rid of --disable-warnings-as-errors on Windows+ARM64 build

### Reviewers
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**)
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/530/head:pull/530`
`$ git checkout pull/530`

### CODE REVIEW

**Comment by lewurm** [comment]
Thank you Andrew for your comments!

> _Mailing list message from [Andrew Haley](mailto:aph@redhat.com) on [hotspot-dev](mailto:hotspot-dev@openjdk.java.net):_
> IMO this warning:
>
> warning C4146: unary minus operator applied to unsigned type, result still unsigned
>
> should not be used.

Okay, added to the Makefile and reverted those changes.

> // Generate stack overflow check
> if (UseStackBanging) {
> - __ bang_stack_with_offset(JavaThread::stack_shadow_zone_size());
> + __ bang_stack_with_offset((int)JavaThread::stack_shadow_zone_size());
> } else {
> Unimplemented();
> 
> Could this one be fixed by changing stack_shadow_zone_size() or
> bang_stack_with_offset() ? I would have thought that whatever type
> stack_shadow_zone_size() returns should be compatible with
> bang_stack_with_offset().

The x86_64 backend and others do the same:
https://github.com/openjdk/jdk/blob/5351ba6cfa8078f503f1cf0c375b692905c607ff/src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp#L2176-L2178

So should we (1) do the same, (2) diverge or (3) fix all of them?



For the remaining comments, I've updated the PR, please have another look.

File: make/autoconf/flags-cflags.m4 [hunk 1]
```diff
@@ -134,6 +134,7 @@ AC_DEFUN([FLAGS_SETUP_WARNINGS],
 
       WARNINGS_ENABLE_ALL="-W3"
       DISABLED_WARNINGS="4800"
+      DISABLED_WARNINGS+=" 4146" # unary minus operator applied to unsigned type, result still unsigned
```
Comment (magicus) [inline make/autoconf/flags-cflags.m4:137]: This change will affect *all* JDK code. I'm not sure this was intended?

If it was intended, I think you need to motivate this more explicitly.

If you only wanted to disable the warning for hotspot, the proper solution would be to add it to DISABLED_WARNINGS_microsoft in make/hotspot/lib/CompileJvm.gmk.

**Review by magicus — CHANGES_REQUESTED** [review]

File: make/autoconf/flags-cflags.m4 (hunk 1 above)
Comment (lewurm) [reply make/autoconf/flags-cflags.m4]: Thank you @magicus! It was indeed meant only for the hotspot part.

**Comment by lewurm** [comment]
@theRealAph I prototyped changing the argument of `bang_stack_with_offset()` from `int` to `size_t` here: https://github.com/lewurm/openjdk/commit/85a8f655aa0cb69ef13a2de44dd64c60caf19852. In that approach casting is essentially pushed down to `bang_stack_with_offset` because the assembler instruction of most (all) architectures that is eventually consuming that offset needs a signed integer anyway. Doesn't seem like a win to me to be honest.

I would rather prefer to go with what we have in this patch (similar to what x86 is doing today):
```patch
--- a/src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp
+++ b/src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp
@@ -1524,7 +1524,7 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,

   // Generate stack overflow check
   if (UseStackBanging) {
-    __ bang_stack_with_offset(JavaThread::stack_shadow_zone_size());
+    __ bang_stack_with_offset((int)JavaThread::stack_shadow_zone_size());
   } else {
     Unimplemented();
   }
```
and leave it with that. What do you think?

**Comment by theRealAph** [comment]
> @theRealAph I prototyped changing the argument of `bang_stack_with_offset()` from `int` to `size_t` here: [lewurm@85a8f65](https://github.com/lewurm/openjdk/commit/85a8f655aa0cb69ef13a2de44dd64c60caf19852). In that approach casting is essentially pushed down to `bang_stack_with_offset` because the assembler instruction of most (all) architectures that is eventually consuming that offset needs a signed integer anyway. Doesn't seem like a win to me to be honest.
> 
> I would rather prefer to go with what we have in this patch (similar to what x86 is doing today):
> 
> ```diff
> --- a/src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp
> +++ b/src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp
> @@ -1524,7 +1524,7 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
> 
>    // Generate stack overflow check
>    if (UseStackBanging) {
> -    __ bang_stack_with_offset(JavaThread::stack_shadow_zone_size());
> +    __ bang_stack_with_offset((int)JavaThread::stack_shadow_zone_size());
>    } else {
>      Unimplemented();
>    }
> ```
> 
> and leave it with that. What do you think?

Fine, but please assert `JavaThread::stack_shadow_zone_size() == (int)JavaThread::stack_shadow_zone_size()`.

If all this sounds a bit paranoid, that's because I am.

Adding casts to shut up compilers is a very risky business, because often (if not in this case) the programmer doesn't understand the code well, and just sprinkles casts everywhere. But those casts disable compile-time type checking, and this leads to risks for future maintainability.

I wonder if we should fix this in a better way, and use this in the future:

```
template <typename T1, typename T2>
T1 checked_cast(T2 thing) {
  guarantee(static_cast<T1>(thing) == thing, "must be");
  return static_cast<T1>(thing);
}
```
Then I promise we'll never need to have this conversation again.

File: src/hotspot/cpu/aarch64/nativeInst_aarch64.hpp [hunk 2]
```diff
@@ -655,7 +655,7 @@ class NativeLdSt : public NativeInstruction {
       return 0;
     }
   }
-  size_t size_in_bytes() { return 1 << size(); }
+  size_t size_in_bytes() { return 1ull << size(); }
```
Comment (stooart-mon) [inline src/hotspot/cpu/aarch64/nativeInst_aarch64.hpp:658]: Capital ULL - I find that easer to search for and it is more consistent.

**Comment by lewurm** [comment]
> Fine, but please assert JavaThread::stack_shadow_zone_size() == (int)JavaThread::stack_shadow_zone_size().

Done.

> Adding casts to shut up compilers is a very risky business, because often (if not in this case) the programmer doesn't understand the code well, and sprinkles casts everywhere. But casts disable compile-time type checking, and this leads to risks for future maintainability.

Full ACK and I appreciate your comments on this!


> I wonder if we should fix it in a better way, and use something like
> this in the future for all compiler warnings:
> 
> ```
> template <typename T1, typename T2>
> T1 checked_cast(T2 thing) {
>   T1 result = static_cast<T1>(thing);
>   guarantee(static_cast<T2>(result) == thing, "must be");
>   return result;
> }
> ```
> 
> I know this is additional work, but I promise we'll never need to have this conversation again.

This sounds like a great idea to me. I assume it doesn't fit into the scope of this PR, therefore I've created [JDK-8254856](https://bugs.openjdk.java.net/browse/JDK-8254856) to track it.

File: src/hotspot/cpu/aarch64/nativeInst_aarch64.hpp (hunk 2 above)
Comment (lewurm) [reply src/hotspot/cpu/aarch64/nativeInst_aarch64.hpp]: Thank you! Fixed.

**Review by magicus — APPROVED** [review]
Build changes look fine now.

**Comment by lewurm** [comment]
@theRealAph does the PR look okay to you now?

**Review by theRealAph — APPROVED** [review]

**Comment by lewurm** [comment]
/integrate

Thank you for the reviews, Magnus and Andrew!

**Comment by lewurm** [comment]
Would you mind sponsor it @theRealAph or @magicus?

**Comment by theRealAph** [comment]
> Would you mind to sponsor it @theRealAph or @magicus?

Hmm, I think you have to integrate it first.
https://wiki.openjdk.java.net/display/SKARA/Pull+Request+Commands#PullRequestCommands-/sponsor

**Comment by lewurm** [comment]
Thank you Andrew.

**Comment by magicus** [comment]
@lewurm 
This patch seems to break on linux-aarch64 with gcc:
open/src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:1501:52: error: comparison of integer expressions of different signedness: 'size_t' {aka 'long unsigned int'} and 'int' [-Werror=sign-compare]
[2020-11-02T13:53:43,298Z]  1501 |     assert(StackOverflow::stack_shadow_zone_size() == (int)StackOverflow::stack_shadow_zone_size(), "must be same");

**Comment by magicus** [comment]
Did you test building this on linux?

**Comment by lewurm** [comment]
@magicus I did test the initial version of this PR on linux+arm64, but not the latest iteration. sorry about that 🙁 

What is the policy here? Submit a revert right away or investigate a fix?

**Comment by magicus** [comment]
@lewurm Open a new JBS issue with the bug. If you can find a fix in a short amount of time (which I would believe should be possible; probably just need a proper cast) it's acceptable to fix it directly. What amounts to a "short amount of time" is left to reasonable judgement; minutes to hours are okay, days are not.

Otherwise, create an anti-delta (revert changeset) to back out your changes, and open yet another JBS issue for re-implementing them correctly.  

 In this case, an alternative short-term fix could also be to remove the assert instead of backing out the entire patch.

**Comment by lewurm** [comment]
https://github.com/openjdk/jdk/pull/1013

**Comment by theRealAph** [comment]
> @lewurm
> This patch seems to break on linux-aarch64 with gcc:

Builds cleanly on Linux/GCC or me.

**Comment by lewurm** [comment]
@theRealAph what gcc version?

I can reproduce with
```console
$ gcc --version
gcc (Ubuntu 9.2.1-9ubuntu2) 9.2.1 20191008
```
which ships in Ubuntu 19.10 as default

**Pull request closed without merging**